toolchain/lower/file_context.cpp

```diff
@@ -288,19 +290,47 @@ auto FileContext::ProcessSpecificEquivalence(
   if (canon_id1.index >= canon_id2.index) {
     // Prefer the earlier index for canonical values.
     std::swap(canon_id1, canon_id2);
-    std::swap(replaced_before1, replaced_before2);
   }
 
   // Update equivalent_specifics_ for all. This is used as an indicator that
   // this specific_id may be the canonical one when reducing the equivalence
   // chains in `IsKnownEquivalence`.
   equivalent_specifics_.Set(specific_id1, canon_id1);
   equivalent_specifics_.Set(specific_id2, canon_id1);
-  specific_functions_.Get(canon_id2)->replaceAllUsesWith(
-      specific_functions_.Get(canon_id1));
-  if (!replaced_before2) {
-    specifics_to_delete.push_back(canon_id2);
+  equivalent_specifics_.Set(canon_id1, canon_id1);
+  equivalent_specifics_.Set(canon_id2, canon_id1);
+}
+
+auto FileContext::UpdateEquivalentSpecific(SemIR::SpecificId specific_id)
+    -> void {
+  if (!equivalent_specifics_.Get(specific_id).has_value()) {
+    return;
+  }
+
+  llvm::SmallVector<SemIR::SpecificId> stack;
+  SemIR::SpecificId specific_to_update = specific_id;
+  while (equivalent_specifics_.Get(
+             equivalent_specifics_.Get(specific_to_update)) !=
+         equivalent_specifics_.Get(specific_to_update)) {
+    stack.push_back(specific_to_update);
+    specific_to_update = equivalent_specifics_.Get(specific_to_update);
```

Comment:
Might be nice to avoid the repeated `equivalent_specifics_.Get(specific_to_update)` here -- maybe a straight `while` loop isn't a great fit?

Reply:
Updated to use local variables, reducing the number of `Get` calls.